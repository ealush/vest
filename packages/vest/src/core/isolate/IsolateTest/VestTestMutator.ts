import { IsolateInspector, IsolateMutator } from 'vestjs-runtime';

import { TIsolateTest } from 'IsolateTest';
import {
  TestAction,
  TestStateMachineAction,
  TestStatus,
  createTestStateMachine,
} from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { VestTestInspector } from 'VestTestInspector';

const TestStateMachine = createTestStateMachine();

export class VestTestMutator {
  static setPending(test: TIsolateTest) {
    VestTestMutator.setStatus(test, TestStatus.PENDING);
  }

  static fail(test: TIsolateTest): void {
    VestTestMutator.setStatus(
      test,
      VestTestInspector.warns(test) ? TestStatus.WARNING : TestStatus.FAILED
    );
  }

  static pass(test: TIsolateTest): void {
    VestTestMutator.setStatus(test, TestStatus.PASSING);
  }

  static warn(test: TIsolateTest): void {
    IsolateMutator.setData(test, 'severity', TestSeverity.Warning);
  }

  static skip(test: TIsolateTest, force?: boolean): void {
    // Without this force flag, the test will be marked as skipped even if it is pending.
    // This means that it will not be counted in "allIncomplete" and its done callbacks
    // will not be called, or will be called prematurely.
    // What this mostly say is that when we have a pending test for one field, and we then
    // start typing in a different field - the pending test will be canceled, which
    // is usually an unwanted behavior.
    // The only scenario in which we DO want to cancel the async test regardless
    // is when we specifically skip a test with `skipWhen`, which is handled by the
    // "force" boolean flag.
    // I am not a fan of this flag, but it gets the job done.
    VestTestMutator.setStatus(test, TestStatus.SKIPPED, force);
  }

  static cancel(test: TIsolateTest): void {
    VestTestMutator.setStatus(test, TestStatus.CANCELED);
  }

  static omit(test: TIsolateTest): void {
    VestTestMutator.setStatus(test, TestStatus.OMITTED);
  }

  static reset(test: TIsolateTest): void {
    VestTestMutator.setStatus(test, TestAction.RESET);
  }

  static setStatus(
    test: TIsolateTest,
    status: TestStateMachineAction,
    payload?: any
  ): void {
    IsolateMutator.setData(
      test,
      'status',
      TestStateMachine.staticTransition(
        IsolateInspector.getData(test).status,
        status,
        payload
      )
    );
  }
}
