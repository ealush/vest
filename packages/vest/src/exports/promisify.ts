import { ErrorStrings } from 'ErrorStrings';
import { invariant, isFunction } from 'vest-utils';

import {
  SuiteResult,
  SuiteRunResult,
  TFieldName,
  TGroupName,
} from 'SuiteResultTypes';

function promisify<F extends TFieldName, G extends TGroupName>(
  validatorFn: (...args: any[]) => SuiteRunResult<F, G>
) {
  return (...args: any[]): Promise<SuiteResult<F, G>> => {
    invariant(isFunction(validatorFn), ErrorStrings.PROMISIFY_REQUIRE_FUNCTION);

    return new Promise(resolve => validatorFn(...args).done(resolve));
  };
}

export default promisify;
