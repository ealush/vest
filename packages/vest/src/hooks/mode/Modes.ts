export enum Modes {
  ALL,
  EAGER,
}
