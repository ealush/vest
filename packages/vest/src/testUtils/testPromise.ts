export function TestPromise(cb: (done: () => void) => void): Promise<void> {
  return new Promise<void>(done => cb(done));
}
