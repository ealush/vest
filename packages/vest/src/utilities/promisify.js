const promisify = validatorFn => (...args) => {
  if (typeof validatorFn !== 'function') {
    throw new Error('[vest/promisify]: Expected validatorFn to be a function.');
  }

  return new Promise(resolve => validatorFn(...args).done(resolve));
};

export default promisify;
