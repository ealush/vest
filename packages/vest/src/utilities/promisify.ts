import isFunction from 'isFunction';

import { IVestResult } from 'produce';
import { TDraftResult } from 'produceDraft';

const promisify =
  (validatorFn: (...args: any[]) => IVestResult) =>
  (...args: any[]): Promise<TDraftResult> => {
    if (!isFunction(validatorFn)) {
      throw new Error(
        '[vest/promisify]: Expected validatorFn to be a function.'
      );
    }

    return new Promise(resolve => validatorFn(...args).done(resolve));
  };

export default promisify;
