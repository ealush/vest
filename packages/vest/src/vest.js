import create from 'createSuite';
import enforce from 'enforce';
import { only, skip, warn, group } from 'hooks';
import test from 'test';

const VERSION = __LIB_VERSION__;

export default {
  VERSION,
  create,
  enforce,
  group,
  only,
  skip,
  test,
  warn,
};
