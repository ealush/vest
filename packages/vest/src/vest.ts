import { enforce } from 'n4s';

import create from 'create';
import context from 'ctx';
import each from 'each';
import { only, skip } from 'exclusive';
import group from 'group';
import include from 'include';
import { eager } from 'mode';
import omitWhen from 'omitWhen';
import optional from 'optionalTests';
import skipWhen from 'skipWhen';
import { test } from 'test';
import warn from 'warn';

const VERSION = __LIB_VERSION__;
export {
  test,
  create,
  each,
  only,
  skip,
  warn,
  group,
  optional,
  skipWhen,
  omitWhen,
  enforce,
  VERSION,
  context,
  include,
  eager,
};
