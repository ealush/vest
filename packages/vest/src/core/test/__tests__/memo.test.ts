import wait from 'wait';

import { TestPromise } from '../../../testUtils/testPromise';

import { Modes } from 'Modes';
import { VestTestMutator } from 'VestTestMutator';
import promisify from 'promisify';
import * as vest from 'vest';
import { test as vestTest, enforce } from 'vest';

describe('test.memo', () => {
  describe('cache hit', () => {
    it('Should return without calling callback', () => {
      const cb1 = jest.fn();
      const cb2 = jest.fn(() => TestPromise(() => undefined));
      const suite = vest.create(() => {
        vestTest.memo('f1', cb1, [1]);
        vestTest.memo('f1', cb2, [2]);
      });

      suite();
      expect(cb1).toHaveBeenCalledTimes(1);
      expect(cb2).toHaveBeenCalledTimes(1);
      suite();
      expect(cb1).toHaveBeenCalledTimes(1);
      expect(cb2).toHaveBeenCalledTimes(1);
    });

    it('Should produce correct initial result', () => {
      const res = vest.create(() => {
        vest.mode(Modes.ALL);
        vestTest.memo('field1', 'msg1', () => false, [{}]);
        vestTest.memo('field1', 'msg2', () => undefined, [{}]);
        vestTest.memo('field2', () => undefined, [{}]);
        vestTest.memo(
          'field3',
          () => {
            vest.warn();
            return false;
          },
          [{}]
        );
      })();

      expect(res.hasErrors('field1')).toBe(true);
      expect(res.hasErrors('field2')).toBe(false);
      expect(res.hasWarnings('field3')).toBe(true);
      expect(res).toMatchSnapshot();
    });
    describe('sync', () => {
      it('Should restore previous result on re-run', () => {
        const suite = vest.create(() => {
          vest.mode(Modes.ALL);
          vestTest.memo('field1', 'msg1', () => false, [1]);
          vestTest.memo('field1', 'msg2', () => undefined, [2]);
          vestTest.memo('field2', () => undefined, [3]);
          vestTest.memo(
            'field3',
            () => {
              vest.warn();
              return false;
            },
            [4]
          );
        });

        const res = suite();

        expect(res.hasErrors('field1')).toBe(true);
        expect(res.hasErrors('field2')).toBe(false);
        expect(res.hasWarnings('field3')).toBe(true);
        expect(res).toMatchSnapshot();

        const res2 = suite();
        expect(res2.hasErrors('field1')).toBe(true);
        expect(res2.hasErrors('field2')).toBe(false);
        expect(res2.hasWarnings('field3')).toBe(true);
        expect(res).isDeepCopyOf(res2);
      });
    });

    describe('async', () => {
      it('Should immediately return previous result on re-run', async () => {
        {
          const suite = vest.create(() => {
            vestTest.memo(
              'field1',
              async () => {
                await wait(500);
                enforce(1).equals(2);
              },
              [1]
            );
            vestTest.memo(
              'field2',
              async () => {
                await wait(500);
                enforce(1).equals(2);
              },
              [2]
            );
          });

          const asyncSuite = promisify(suite);

          let start = Date.now();
          const res1 = await asyncSuite();
          enforce(Date.now() - start).gte(500);

          start = Date.now();
          const res2 = suite();

          expect(res1).isDeepCopyOf(res2);
        }
      });
    });

    describe('Test is canceled', () => {
      it('Should refresh', async () => {
        let count = 0;
        const tests: vest.IsolateTest[] = [];
        const suite = vest.create(() => {
          count++;

          tests.push(
            vestTest.memo(
              'f1',
              async () => {
                await wait(10);
              },
              [true]
            )
          );

          if (count === 1) {
            VestTestMutator.cancel(tests[0]);
          }
        });

        suite();
        suite();
        suite();

        expect(tests[0]).not.toBe(tests[1]);
        expect(tests[1]).toBe(tests[2]);
      });
    });
  });

  describe('cache miss', () => {
    it('Should run test normally', () => {
      const cb1 = jest.fn(res => res);
      const cb2 = jest.fn(
        res => new Promise<void>((resolve, rej) => (res ? resolve() : rej()))
      );
      const suite = vest.create((key, res) => {
        vestTest.memo('f1', () => cb1(res), [1, key]);
        vestTest.memo('f2', () => cb2(res), [2, key]);
      });

      expect(cb1).toHaveBeenCalledTimes(0);
      expect(cb2).toHaveBeenCalledTimes(0);
      suite('a', false);
      expect(cb1).toHaveBeenCalledTimes(1);
      expect(cb2).toHaveBeenCalledTimes(1);
      expect(suite.get().hasErrors()).toBe(true);
      suite('b', true);
      expect(cb1).toHaveBeenCalledTimes(2);
      expect(cb2).toHaveBeenCalledTimes(2);
      expect(suite.get().hasErrors()).toBe(false);
    });
  });

  describe('Collision detection', () => {
    describe('cross-field collision', () => {
      it('Should factor in field name', () => {
        const suite = vest.create(() => {
          vestTest.memo('f1', () => false, [1]);
          vestTest.memo('f2', () => true, [1]);
        });

        suite();
        suite();
        expect(suite.get().hasErrors('f1')).toBe(true);
        expect(suite.get().hasErrors('f2')).toBe(false);
      });
    });

    describe('same-field-same-suite collision', () => {
      it('Should factor in execution order', () => {
        const suite = vest.create(() => {
          vestTest.memo('f1', () => false, [1]);
          vestTest.memo('f1', () => true, [1]);
        });

        suite();
        suite();
        expect(suite.get().hasErrors('f1')).toBe(true);
        expect(suite.get().errorCount).toBe(1);
      });
    });
    describe('cross-suite collision', () => {
      it('Should factor in field name', () => {
        const suite1 = vest.create(() => {
          vestTest.memo('f1', () => false, [1]);
          vestTest.memo('f2', () => true, [1]);
        });
        const suite2 = vest.create(() => {
          vestTest.memo('f1', () => true, [1]);
          vestTest.memo('f2', () => false, [1]);
        });

        suite1();
        suite2();
        expect(suite1.get().hasErrors('f1')).toBe(true);
        expect(suite1.get().hasErrors('f2')).toBe(false);
        expect(suite2.get().hasErrors('f1')).toBe(false);
        expect(suite2.get().hasErrors('f2')).toBe(true);
      });
    });
  });
});
