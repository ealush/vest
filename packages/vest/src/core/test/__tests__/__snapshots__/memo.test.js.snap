// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test.memo Async tests Cache miss Should run test functions normally 1`] = `
Object {
  "errorCount": 1,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "name": "suite_name",
  "testCount": 1,
  "tests": Object {
    "field_1": Object {
      "errorCount": 1,
      "testCount": 1,
      "warnCount": 0,
    },
  },
  "warnCount": 0,
}
`;

exports[`test.memo Sync tests On cache hit Should produce correct validation result 1`] = `
Object {
  "done": [Function],
  "errorCount": 2,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "name": "suite_name",
  "testCount": 4,
  "tests": Object {
    "field_1": Object {
      "errorCount": 1,
      "errors": Array [
        "message",
      ],
      "testCount": 1,
      "warnCount": 0,
    },
    "field_2": Object {
      "errorCount": 1,
      "errors": Array [
        "message",
      ],
      "testCount": 1,
      "warnCount": 0,
    },
    "field_3": Object {
      "errorCount": 0,
      "testCount": 1,
      "warnCount": 0,
    },
    "field_4": Object {
      "errorCount": 0,
      "testCount": 1,
      "warnCount": 1,
      "warnings": Array [
        "message",
      ],
    },
  },
  "warnCount": 1,
}
`;

exports[`test.memo Sync tests On cache miss Should produce correct validation result 1`] = `
Object {
  "done": [Function],
  "errorCount": 2,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "name": "suite_name",
  "testCount": 4,
  "tests": Object {
    "field_1": Object {
      "errorCount": 1,
      "errors": Array [
        "message",
      ],
      "testCount": 1,
      "warnCount": 0,
    },
    "field_2": Object {
      "errorCount": 1,
      "errors": Array [
        "message",
      ],
      "testCount": 1,
      "warnCount": 0,
    },
    "field_3": Object {
      "errorCount": 0,
      "testCount": 1,
      "warnCount": 0,
    },
    "field_4": Object {
      "errorCount": 0,
      "testCount": 1,
      "warnCount": 1,
      "warnings": Array [
        "message",
      ],
    },
  },
  "warnCount": 1,
}
`;

exports[`test.memo Sync tests On cache update Should produce fresh result 1`] = `
Object {
  "done": [Function],
  "errorCount": 2,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "name": "suite_name",
  "testCount": 4,
  "tests": Object {
    "field_1": Object {
      "errorCount": 1,
      "errors": Array [
        "message",
      ],
      "testCount": 1,
      "warnCount": 0,
    },
    "field_2": Object {
      "errorCount": 0,
      "testCount": 1,
      "warnCount": 0,
    },
    "field_3": Object {
      "errorCount": 1,
      "errors": Array [
        "message",
      ],
      "testCount": 1,
      "warnCount": 0,
    },
    "field_4": Object {
      "errorCount": 0,
      "testCount": 1,
      "warnCount": 0,
    },
  },
  "warnCount": 0,
}
`;
